data_source.go

```diff
@@ -303,39 +299,36 @@ func (ds *pgDataSource) buildConnInfo(conninfo map[string]string) string {
 		ds.cfg.Database, ds.cfg.User, ds.cfg.Sslmode, ds.cfg.BinaryParameters, ds.cfg.Password)
 }
 
-func (ds *pgDataSource) getPgCurrentWalLsn(role string) (string, error) {
-	if role == "replica" {
-		// query select * from pg_stat_wal_receiver;  to get conninfo
-		//create a new db connection to upstream (primary) with conninfo and return pg_current_wal_lsn
-		conninfo, err := ds.getUpstreamConnInfo()
+func (ds *pgDataSource) getPgCurrentWalLsn(maxHop int64, db *sqlx.DB) (string, error) {
+	var isReplica bool
+	err := db.Get(&isReplica, "select pg_catalog.pg_is_in_recovery()")
+	if err != nil {
+		return "", err
+	}
+
+	if isReplica {
+		if maxHop == 0 {
+			return "", errors.New("Reached max hop limit")
+		}
+
+		conninfo, err := ds.getUpstreamConnInfo(db)
 		if err != nil {
 			return "", err
 		}
 		upstreamConnInfo := ds.buildConnInfo(parseConnInfo(conninfo))
 		upstreamDb, err := sqlConnect(upstreamConnInfo)
```

Comment:
Let's check this err != nil here and then we should be set.

Reply:
great catch! fixed!